demo/lib/vm.bash

```diff
@@ -112,6 +123,53 @@ vm-wait-process() {
     fi
 }
 
+vm-copy-file() { # script API
+    # Usage: vm-copy-file PATH-IN-VM
+    #
+    # Reads stdin and writes the content to PATH-IN-VM, creating any
+    # intermediate directories necessary.
+    #
+    # Example:
+    #   echo 'Ahoy, Matey...' | vm-write-file /etc/motd
+    if [ "$1" == "--cleanup" ]; then
+        shift
+        trap "rm -f $1" RETURN EXIT
```

Comment:
First, I also had an attempt at another variant (with an IMO much nicer approach of) `vm-write-file`, which simply assumed the content would come piped through `stdin` and then was trying to write the content to the specified location on the VM, by doing something like this:
```
vm-write-file() {
    local file="$1"; shift
    case "$file" in
        */*) ;;
        *) file="./$file";;
    esac
    vm-command-q "mkdir -p ${file%/*} && cat > $file" ||
        command-error "failed to write to file $file"
}

# this was then supposed to be used like this (I was a it surprised by the correct syntax)
cat <<<EOF |
a template
file
- with $VARIOUS
- $VARIABLES of all $SORTS
- with $(date) thrown in
EOF
    vm-write-file /path/on/VM
```

I guess you guess what happened. Defeated by that  `<<<"$1"` in `vm-command-q`. For a short while I was considering as a workaround adding a `vm-pipe-input()` which would not use the `sudo bash -l <<<"$1"` idiom. But I quickly gave up the idea because I think sooner than later it would lead down the rabbit hole/rathole of getting locked up in the nuthouse because of quoting insanity.